### PR TITLE
Create Canister migration

### PR DESCRIPTION
**Pull request by EsdrasVP**
This PR is able to:
- Create a canister migration
- Add its reference to hardwares table

A Canister is a controller inside the Physical Storage (which may have more than one Canister). It contains its own firmware, ports, network, power state, etc.

Depends on:
- [~ManageIQ/manageiq-schema#233 - Merged~](https://github.com/ManageIQ/manageiq-schema/pull/233)
- [ManageIQ/manageiq-schema#269](https://github.com/ManageIQ/manageiq-schema/pull/269)
- [~ManageIQ/manageiq#17700 - Merged~](https://github.com/ManageIQ/manageiq/pull/17700)

### CODE REVIEW

**Comment by EsdrasVP** [comment]
@miq-bot assign @agrare

**Comment by agrare** [comment]
@EsdrasVP can you add to the PR description what a canister is?


> Create a canister migration

Isn't enough for me to understand what is being added

**Comment by EsdrasVP** [comment]
> @EsdrasVP can you add to the PR description what a canister is?
> 
> Create a canister migration
> 
> Isn't enough for me to understand what is being added

@agrare I updated the description explaining what a Canister is. Can you see if it's more understandable?

**Comment by agrare** [comment]
@EsdrasVP is Canister an industry standard term for this type of object?  (A quick google search for "storage canister" yielded a lot of amazon results for air-tight coffee bean containers which might just be my recommended search results :laughing:)

If it has its own controller it almost sounds like a caniter is its own self-contained storage?

**Comment by EsdrasVP** [comment]
> @EsdrasVP is Canister an industry standard term for this type of object? (A quick google search for "storage canister" yielded a lot of amazon results for air-tight coffee bean containers which might just be my recommended search results )

@agrare Other companies might call it by a different name, so I think that's why it's only appearing as coffee bean containers on search results hahaha.

> If it has its own controller it almost sounds like a caniter is its own self-contained storage?

Yes, all the work will be done by the Canisters, which will exist inside of a PhysicalStorage. This distinction is important because a Lenovo PhysicalStorage will contain more than one Canister inside it.

**Comment by agrare** [comment]
Okay canister seems like a decent enough name don't want to dwell on that.

File: db/migrate/20180713194201_create_canister.rb [hunk 1]
```diff
@@ -0,0 +1,21 @@
+class CreateCanister < ActiveRecord::Migration[5.0]
+  def change
+    create_table :canisters do |t|
+      t.string :serial_number
```
Comment (agrare) [inline db/migrate/20180713194201_create_canister.rb]: 2 things here, can you add ems_ref for a unique reference?  Even if e.g. position is unique for lenovo we can use ems_ref generally.
Also I assume this belongs to a physical_storage but you don't have a reference to physical_storage here.

File: db/migrate/20180730134426_add_port_status_to_physical_network_ports.rb [hunk 2]
```diff
@@ -0,0 +1,5 @@
+class AddPortStatusToPhysicalNetworkPorts < ActiveRecord::Migration[5.0]
+  def change
+    add_column :physical_network_ports, :port_status, :string
```
Comment (agrare) [inline db/migrate/20180730134426_add_port_status_to_physical_network_ports.rb]: This looks unrelated to the rest of this PR

File: db/migrate/20180713194201_create_canister.rb (hunk 1 above)
Comment (EsdrasVP) [reply db/migrate/20180713194201_create_canister.rb]: @agrare I added now `ems_ref` and `physical_storage_id` to the migration.

File: db/migrate/20180730134426_add_port_status_to_physical_network_ports.rb (hunk 2 above)
Comment (EsdrasVP) [reply db/migrate/20180730134426_add_port_status_to_physical_network_ports.rb]: @agrare It is, kinda. Because Canister will have some physical ports in it, and I didn't found any reference to this, so I created it. If you want to, I can put it in another PR.

Comment (agrare) [reply db/migrate/20180730134426_add_port_status_to_physical_network_ports.rb]: Yeah lets do this in a new PR since port_status can apply to anything not just canisters.  So a canister will have hardware -> guest_devices -> physical_network_ports?

Comment (EsdrasVP) [reply db/migrate/20180730134426_add_port_status_to_physical_network_ports.rb]: > So a canister will have hardware -> guest_devices -> physical_network_ports?

@agrare Exactly, this relationship it's currently on the core PR.

Comment (EsdrasVP) [reply db/migrate/20180730134426_add_port_status_to_physical_network_ports.rb]: > Yeah lets do this in a new PR since port_status can apply to anything not just canisters.

@agrare I've done it and put as dependency on this PR, could you take a look there?

**Comment by agrare** [comment]
@EsdrasVP should a physical_disk have a canister_id then?

**Comment by EsdrasVP** [comment]
> @EsdrasVP should a physical_disk have a canister_id then?

@agrare No, because the PhysicalDisks are on the same level as Canisters on a PhysicalStorage. It's quite bad actually, because doesn't allow us to see which canister a disk is inside.
